### PR TITLE
Add Qwen3-VL Multimodal Training Support with vLLM 0.20

### PR DESCRIPTION
**Pull request by shx2005**
## Purpose

Add Qwen3-VL Multimodal Training Support with vLLM 0.20

## Description

This PR adds `prepare_data.py --multimodal` for single-image image-text preprocessing. In multimodal mode, preprocessing uses the model processor to render and tokenize the chat prompt, expands image placeholders with the actual processor image inputs so token lengths match vLLM, keeps the existing training fields (`input_ids`, `loss_mask`, `seq_len`), and preserves normalized `messages` for vLLM chat-based hidden-state generation.

The offline datagen path and online on-missing generation path now carry those `messages` alongside the preprocessed `input_ids`. When `messages` are present, the vLLM client converts processor-style image parts into OpenAI-compatible chat content and sends the request through the chat completions API with `return_token_ids` and `add_generation_prompt=false`. Text-only data remains on the existing token-id completions path.

For vLLM 0.20 chat requests, the returned prompt token IDs can correspond to the fully rendered chat prompt. The multimodal path therefore accepts the preprocessed `input_ids` as a prefix match and trims the saved hidden states back to the preprocessed sequence length.

This PR also includes follow-up fixes found during validation:

- keep text-only preprocessing free of multimodal columns
- strip the verifier final layer from explicit target-layer ids before saving draft auxiliary layer ids
- preserve the configured hidden-state dtype for empty batches created by `--on-missing skip`

This commit ended up larger than expected, and progress was slower than planned, because vLLM 0.20 changed the multimodal message format and surfaced prompt-token inconsistencies during validation. Resolving those mismatches required additional compatibility handling, regression coverage, and a runnable Qwen3-VL 5k online training example. I also validated that the same flow scales from the 5k example dataset to the 48k LLaVA-CoT variant.

## Related Issue

https://github.com/vllm-project/speculators/issues/290

## Tests

Passed:

- `bash examples/train/eagle3_qwen3_vl_4b_llava_cot_5k_online.sh`
- `bash examples/train/eagle3_qwen3_8b_sharegpt_online_5k.sh` (regression test for the existing text-only path)
- `make quality`
- 5k Qwen3-VL online example:

    ` Timing from an observed run on 4x NVIDIA GeForce RTX 5090 32GB GPUs`
    ` (vLLM on GPUs 0,1 and training on GPUs 2,3):`
    ` Data preprocessing: 460 seconds (7 mins 40 secs)`
    ` vLLM server startup: 45 seconds`
    ` Training (5 epochs): 1110 seconds (18 mins 30 secs)`
     ` Total (prepare_data start to checkpoint save): 1615 seconds (26 mins 55 secs)`
image

  
- 48k Qwen3-VL online run (only 1 epoch):
  - `val/loss_epoch=6.894`
  - `val/full_acc_0_epoch=65.5%`
  - `val/full_acc_1_epoch=40.6%`
  - `val/full_acc_2_epoch=25.3%`

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

**Comment by shx2005** [comment]
@coderabbitai review

**Comment by shx2005** [comment]
@shanjiaz I’ve finished the single image-text testing and code cleanup. When you have time, would you mind taking a look? I’d be very grateful for any feedback or suggestions.

The commit ended up being a bit larger than I initially expected, mainly because vLLM 0.20 changed the multimodal message format. To address that, I added some compatibility handling and extra validation to make sure the multimodal path works correctly without affecting the existing text-only path.

Thank you very much for your patience, guidance, and continued support. I really appreciate all your help throughout this process. Please feel free to let me know if anything looks unclear or if there is anything you would like me to adjust.

**Comment by shanjiaz** [comment]
> @shanjiaz I’ve finished the single image-text testing and code cleanup. When you have time, would you mind taking a look? I’d be very grateful for any feedback or suggestions.
> 
> The commit ended up being a bit larger than I initially expected, mainly because vLLM 0.20 changed the multimodal message format. To address that, I added some compatibility handling and extra validation to make sure the multimodal path works correctly without affecting the existing text-only path.
> 
> Thank you very much for your patience, guidance, and continued support. I really appreciate all your help throughout this process. Please feel free to let me know if anything looks unclear or if there is anything you would like me to adjust.

Thank you so much for taking this up! Will take a look. : )
